---
title: "Miracle_Test"
output:
  html_document: default
  pdf_document: default
date: "2024-11-04"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Load packages
```{r}
library(tidyverse)
library(edgeR)
library(ggfortify)
library(Hmisc)
library(corrplot)
library(GGally)
```


#Preparing data
```{r preparing data}
setwd("C:/Users/Gina/Desktop/Test")
lung_data <- read_tsv("Lung.txt",show_col_types=FALSE)
sample_infomation <- read_tsv("Sample_information.txt",show_col_types=FALSE)
lung <- as.matrix(lung_data)
sample_infomation <- as.data.frame(sample_infomation)
```


```{r preparing data}
lung_sample <- lung[,1]
rownames(lung)<-lung_sample
lung <- lung[,-1]
lung_feature <- colnames(lung)
lung<-t(lung)
rownames(sample_infomation) <- sample_infomation[,1]
sample_infomation <- sample_infomation[,-1]
```

 
#Differential Gene Expression Analysis

##1	Generate an edgeR object

```{r convert_to_DGEList}
raw_counts <- lung# a matrix with gene names as row names and sample names as column names
# Now you can assign gene counts matrix and sample information data frame to edgeR object
lung_edgeR <- DGEList(counts=raw_counts,
                      samples=sample_infomation) 
lung_edgeR$samples$group <- lung_edgeR$samples$CON# define a grouping factor
lung_edgeR$samples$stage <- lung_edgeR$samples$STAGE# define a grouping factor

```

##2 normalisation

```{r log2_transformation}
lung_lcpm <- lung_edgeR %>%
  cpm(log=TRUE, prior.count=1)
```

##3 design matrix

###Create design matrix
```{r Create a design matrix}
lung_edgeR$samples$group <- relevel(factor(lung_edgeR$samples$group), ref = "NC")
lung_edgeR$samples$stage <- relevel(factor(lung_edgeR$samples$stage), ref = "NN")
design_stage <- model.matrix(~ stage, lung_edgeR$samples)
design <- model.matrix(~group, lung_edgeR$samples)
```

##4 differential expression analysis

###CON
```{r differential expression analysis}
library(limma)
#
v <- voom(lung_edgeR, design)
fit <- lmFit(v, design)
fit <- eBayes(fit)
results<-topTable(fit, coef=2)
#
v_lcpm <- voom(lung_lcpm, design)
fit_lcpm <- lmFit(v_lcpm, design)
fit_lcpm <- eBayes(fit_lcpm)
results_lcpm<-topTable(fit_lcpm, coef=2)

```


```{r find features}
significant_genes <- results[results$P.Value < 0.05, ]
significant_genes <- rownames(significant_genes)  
significant_genes_lcpm <- results_lcpm[results_lcpm$P.Value < 0.05, ]
significant_genes_lcpm <- rownames(significant_genes_lcpm)  
feature_genes <- intersect(significant_genes, significant_genes_lcpm)
```

###STAGE
```{r}
v_satge <- voom(lung_edgeR, design_stage)
fit_stage <- lmFit(v_satge, design_stage)
fit_stage <- eBayes(fit_stage)
results1<-topTable(fit_stage, coef=2)
results2<-topTable(fit_stage, coef=3)
results3<-topTable(fit_stage, coef=4)
results4<-topTable(fit_stage, coef=5)
results5<-topTable(fit_stage, coef=6)
results6<-topTable(fit_stage, coef=7)
results7<-topTable(fit_stage, coef=8)


v_lcpm_stage <- voom(lung_lcpm, design_stage)
fit_lcpm_stage <- lmFit(v_lcpm_stage, design_stage)
fit_lcpm_stage <- eBayes(fit_lcpm_stage)
results_lcpm1<-topTable(fit_lcpm_stage, coef=2)
results_lcpm2<-topTable(fit_lcpm_stage, coef=3)
results_lcpm3<-topTable(fit_lcpm_stage, coef=4)
results_lcpm4<-topTable(fit_lcpm_stage, coef=5)
results_lcpm5<-topTable(fit_lcpm_stage, coef=6)
results_lcpm6<-topTable(fit_lcpm_stage, coef=7)
results_lcpm7<-topTable(fit_lcpm_stage, coef=8)

significant_genes1 <- results1[results1$P.Value < 0.05, ]
significant_genes2 <- results2[results2$P.Value < 0.05, ]
significant_genes3 <- results3[results3$P.Value < 0.05, ]
significant_genes4 <- results4[results4$P.Value < 0.05, ]
significant_genes5 <- results5[results5$P.Value < 0.05, ]
significant_genes6 <- results6[results6$P.Value < 0.05, ]
significant_genes7 <- results7[results7$P.Value < 0.05, ]

significant_genes1 <- rownames(significant_genes1)
significant_genes2 <- rownames(significant_genes2)
significant_genes3 <- rownames(significant_genes3) 
significant_genes4 <- rownames(significant_genes4) 
significant_genes5 <- rownames(significant_genes5) 
significant_genes6 <- rownames(significant_genes6) 
significant_genes7 <- rownames(significant_genes7) 

significant_genes_lcpm1 <- results_lcpm1[results_lcpm1$P.Value < 0.05, ]
significant_genes_lcpm2 <- results_lcpm2[results_lcpm2$P.Value < 0.05, ]
significant_genes_lcpm3 <- results_lcpm3[results_lcpm3$P.Value < 0.05, ]
significant_genes_lcpm4 <- results_lcpm4[results_lcpm4$P.Value < 0.05, ]
significant_genes_lcpm5 <- results_lcpm5[results_lcpm5$P.Value < 0.05, ]
significant_genes_lcpm6 <- results_lcpm6[results_lcpm6$P.Value < 0.05, ]
significant_genes_lcpm7 <- results_lcpm7[results_lcpm7$P.Value < 0.05, ]


significant_genes_lcpm1 <- rownames(significant_genes_lcpm1)
significant_genes_lcpm2 <- rownames(significant_genes_lcpm2)
significant_genes_lcpm3 <- rownames(significant_genes_lcpm3)  
significant_genes_lcpm4 <- rownames(significant_genes_lcpm4)  
significant_genes_lcpm5 <- rownames(significant_genes_lcpm5) 
significant_genes_lcpm6 <- rownames(significant_genes_lcpm6)  
significant_genes_lcpm7 <- rownames(significant_genes_lcpm7)  

feature_genes1 <- intersect(significant_genes1, significant_genes_lcpm1)
feature_genes2 <- intersect(significant_genes2, significant_genes_lcpm2)
feature_genes3 <- intersect(significant_genes3, significant_genes_lcpm3)
feature_genes4 <- intersect(significant_genes4, significant_genes_lcpm4)
feature_genes5 <- intersect(significant_genes5, significant_genes_lcpm5)
feature_genes6 <- intersect(significant_genes6, significant_genes_lcpm6)
feature_genes7 <- intersect(significant_genes7, significant_genes_lcpm7)

feature_lists_lc <- list(feature_genes1, feature_genes2, feature_genes3, feature_genes4)
gene_counts_lc <- table(unlist(feature_lists_lc))
relaxed_features_lc <- names(gene_counts_lc[gene_counts_lc >= 3])

feature_lists_nc <- list(feature_genes5, feature_genes6, feature_genes7)
gene_counts_nc <- table(unlist(feature_lists_nc))
relaxed_features_nc <- names(gene_counts_nc[gene_counts_nc >= 2])

features_lcstage <- Reduce(intersect, list(feature_genes1, feature_genes2, feature_genes3, feature_genes4))
features_ncstage <- Reduce(intersect, list(feature_genes5, feature_genes6, feature_genes7))

features_stage <- intersect(relaxed_features_nc, relaxed_features_lc)
features_stage <- union(feature_stage, features_ncstage)
features_stage <- union(feature_stage, features_lcstage)

```

##4 Summarise
```{r Summarise genes number of down- and up-regulated}
summary(decideTests(fit))
summary(decideTests(fit_stage))
```

##5 Prepare Training Data 
```{r}
genes <- lung_edgeR[feature_genes,] 
genes_lcpm<-lung_lcpm[feature_genes,] 
genes_stage <- lung_edgeR[features_stage,] 
genes_lcpm_stage<-lung_lcpm[features_stage,]

expression_data <- as.data.frame(t(genes_lcpm))
expression_data$group <- lung_edgeR$samples$group
write.csv(expression_data, file = "expression_data.csv",row.names = FALSE)

expression_data_stage <- as.data.frame(t(genes_lcpm_stage))
expression_data_stage$stage <- lung_edgeR$samples$STAGE
write.csv(expression_data_stage, file = "expression_data_stage.csv",row.names = FALSE)
```



